jib-core/src/main/java/com/google/cloud/tools/jib/registry/BlobPuller.java

```diff
@@ -42,18 +45,31 @@
    */
   private final OutputStream destinationOutputStream;
 
+  // TODO: Refactor into BlobPullMonitor.
+  private final Consumer<Long> sizeConsumer;
+  private final Consumer<Long> receivedByteCountConsumer;
+  private final Duration delayBetweenCallbacks = Duration.ofMillis(100);
+
   BlobPuller(
       RegistryEndpointRequestProperties registryEndpointRequestProperties,
       DescriptorDigest blobDigest,
-      OutputStream destinationOutputStream) {
+      OutputStream destinationOutputStream,
+      Consumer<Long> sizeConsumer,
+      Consumer<Long> receivedByteCountConsumer) {
     this.registryEndpointRequestProperties = registryEndpointRequestProperties;
     this.blobDigest = blobDigest;
     this.destinationOutputStream = destinationOutputStream;
+    this.sizeConsumer = sizeConsumer;
+    this.receivedByteCountConsumer = receivedByteCountConsumer;
   }
 
   @Override
   public Void handleResponse(Response response) throws IOException, UnexpectedBlobDigestException {
-    try (OutputStream outputStream = destinationOutputStream) {
+    sizeConsumer.accept(response.getContentLength());
+
+    try (OutputStream outputStream =
+        new ListenableCountingOutputStream(
+            destinationOutputStream, receivedByteCountConsumer, delayBetweenCallbacks)) {
       BlobDescriptor receivedBlobDescriptor = response.getBody().writeTo(outputStream);
```

Comment:
For `BlobHttpContent`, you create `new ListenableCountingOutputStream` inside `writeTo()`. It may just be that the names are idential, but does it make sense to create the stream inside `Blob.writeTo()` similarly? (Probably not.)